app/packages/core/src/components/Filters/categoricalFilter/filterOption/FilterOption.tsx

```diff
@@ -0,0 +1,301 @@
+  shouldNotShowExclude: boolean; // for BooleanFields
+  excludeAtom: RecoilState<boolean>;
+  onlyMatchAtom: RecoilState<boolean>;
+  isMatchingAtom: RecoilState<boolean>;
+  valueName: string;
+  color: string;
+  modal: boolean;
+  isKeyPointLabel: boolean;
+}
+
+type Option = {
+  key: string;
+  value: string;
+  icon?: string;
+  tooltip: string;
+};
+
+type Key = "filter" | "negativeFilter" | "match" | "negativeMatch";
+
+const generateOptions = (
+  nestedField: string | undefined,
+  shouldNotShowExclude: boolean,
+  modal: boolean,
+  isKeyPointLabel: boolean,
+  valueName: string
+) => {
+  //  feature requirements:
+  //  1) only nested ListField items should have the filter and negative filter options;
+  //  2) BooleanField should not have the negative filter or negative match options;
+  //  3) in expanded mode or keypoints field, do not show the match or negative match options;
+  const options: Option[] = [];
+  if (Boolean(nestedField)) {
+    options.push({
+      icon: "FilterAltIcon",
+      key: "filter",
+      value: `Select ${nestedField} with ${valueName}`,
+      tooltip: "dataset.filter_labels(field, expr, only_matches=True)",
+    });
+  }
+  if (Boolean(nestedField) && !shouldNotShowExclude) {
+    options.push({
+      icon: "FilterAltOffIcon",
+      key: "negativeFilter",
+      value: `Exclude ${nestedField} with ${valueName}`,
+      tooltip: "dataset.filter_labels(field, ~expr, only_matches=False)",
+    });
+  }
+  if (!modal && !isKeyPointLabel) {
+    options.push({
+      icon: "ImageIcon",
+      key: "match",
+      value: `Show samples with ${valueName}`,
+      tooltip: Boolean(nestedField)
+        ? "dataset.match_labels(fields=field, filter=expr)"
+        : "dataset.match(expr)",
+    });
+  }
+  if (!modal && !shouldNotShowExclude && !isKeyPointLabel) {
+    options.push({
+      icon: "HideImageIcon",
+      key: "negativeMatch",
+      value: `Omit samples with ${valueName}`,
+      tooltip: Boolean(nestedField)
+        ? "dataset.match_labels(fields=field, filter=expr, bool=False)"
+        : "dataset.match(~expr)",
+    });
+  }
+  return options;
+};
+
+const Text = styled.div`
+  font-size: 1rem;
+  margin: auto auto auto 5px;
+  ${({ theme }) => theme.text.secondary};
+`;
+
+const FilterOptionContainer = styled.div`
+  position: relative;
+`;
+
+const FilterOption: React.FC<Props> = ({
+  color,
+  modal,
+  isKeyPointLabel,
+  valueName,
+  nestedField,
+  shouldNotShowExclude,
+  excludeAtom,
+  onlyMatchAtom,
+  isMatchingAtom,
+}) => {
+  const [key, setKey] = React.useState<Key | null>(null);
+
+  const [open, setOpen] = React.useState(false);
+  const [excluded, setExcluded] = useRecoilState(excludeAtom);
+  const [onlyMatch, setOnlyMatch] = useRecoilState(onlyMatchAtom);
+  const [isMatching, setIsMatching] = useRecoilState(isMatchingAtom);
+
+  const theme = useTheme();
+  const highlightedBGColor = Color(color).alpha(0.25).string();
+
+  const popoutRef = React.useRef();
```

Comment:
Refs should be explicitly typed when they are initially undefined

```suggestion
  const popoutRef = React.useRef<HTMLDivElement>();
```